Add sql package and common adapter

A9u opened the pull request.

or-else commented:
Hi. I don't fully understand the purpose of this PR. This PR contains a copy of mysql adapter under a different name. It still contains MySQL-specific code. It won't work for any other SQL database.

It's fine to create a common SQL adapter but it should be generic. 

Also, when sending a PR please send it to `devel` branch. Thanks!

A9u commented:
@or-else Hi, I have started working on the Postgres adapter. I have created a draft PR (work in progress) so that you can see my changes and suggest me changes as I am progressing. When it is ready, I will mark it for review. 
Following are the changes that I have done:- 
 + I have created a sql package. 
 + I have copied the adapter.go file from mysql to sql package.  
 + I have added a driver field to the adapter struct. 
 + In the adapter.go file, upto SetMaxResults function, I have removed mysql and instead used driver name of the struct. 

I am stuck  and was wondering if you could suggest me how to proceed further. 
The CreateDb function creates a database with character set and collate set. 
For Postgres, the syntax is different than Mysql. 
So I was not sure how to proceed further. Can you please help me here.

or-else commented:
There are two options:
1. Create a PgSQL-specific adapter. In other words, target https://github.com/go-pg/pg or https://github.com/lib/pq, use PgSQL-specific syntax for queries etc.

2. Try to create a generic SQL adapter. In that case target https://golang.org/pkg/database/sql/ and write queries which platform-independent.

I think 2 is much harder than 1. Also, I don't think the complexity of 2 is justified. MySQL + PgSQL cover 90% of open source SQL DBs (the remaining 10% is SQLite, which is not relevant to this use case).  

I would strongly suggest to go with 1: pick one of the two main PgSQL drivers go-pg/pg or lib/pq and write specifically to them. The second option is just going to be much harder and not that useful.

A9u commented:
@or-else  Thanks a lot for the suggestion. I will look into it.

or-else commented:
Hi. This seems to be abandoned. Do you have plans to continue or should I just close it?

A9u commented:
Hi, I am working on it. I got little busy on other items. I will pick this up in a few days time.

or-else commented:
That's great, thanks! 

You may want to periodically push changes to this PR.  Take a look how Mongo adapter is being developed:
https://github.com/tinode/chat/pull/306

or-else commented:
Hi @A9u !

It's been two months without changes. Is it moving forward or should I close this PR?

Thanks.

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+func (a *adapter) UserUnreadCount(uid t.Uid) (int, error) {
+	var count int
+	err := a.db.Get(&count, "SELECT SUM(t.seqid)-SUM(s.readseqid) FROM topics AS t, subscriptions AS s "+
+		"WHERE s.userid=? AND t.name=s.topic AND s.deletedat IS NULL AND t.deletedat IS NULL AND "+
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+func (a *adapter) topicCreate(tx *sqlx.Tx, topic *t.Topic) error {
+	_, err := tx.Exec("INSERT INTO topics(createdAt,updatedAt,touchedAt,name,owner,access,public,tags) "+
+		"VALUES(?,?,?,?,?,?,?,?)",
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	decoded_uid := store.DecodeUid(t.ParseUid(sub.User))
+	_, err := tx.Exec(
+		"INSERT INTO subscriptions(createdAt,updatedAt,deletedAt,userid,topic,modeWant,modeGiven,private) "+
+			"VALUES(?,?,NULL,?,?,?,?,?)",
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+	if err != nil && isDupe(err) {
+		if undelete {
+			_, err = tx.Exec("UPDATE subscriptions SET createdAt=?,updatedAt=?,deletedAt=NULL,modeGiven=? "+
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+		} else {
+			_, err = tx.Exec(
+				"UPDATE subscriptions SET createdAt=?,updatedAt=?,deletedAt=NULL,modeWant=?,modeGiven=?,private=? "+
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	// Fetch topic by name
+	var tt = new(t.Topic)
+	err := a.db.Get(tt,
+		"SELECT createdat,updatedat,deletedat,touchedat,name AS id,access,owner,seqid,delid,public,tags FROM topics WHERE name=?",
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+// Reads and denormalizes Public value.
+func (a *adapter) TopicsForUser(uid t.Uid, keepDeleted bool, opts *t.QueryOpt) ([]t.Subscription, error) {
+	// Fetch user's subscriptions
+	q := `SELECT createdat,updatedat,deletedat,topic,delid,recvseqid,
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	// Fetch p2p users and join to p2p tables
+	if err == nil && len(usrq) > 0 {
+		q, usrq, _ := sqlx.In(
+			"SELECT id,state,createdat,updatedat,deletedat,access,lastseen,useragent,public,tags FROM users WHERE id IN (?)",
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	q := `SELECT s.createdat,s.updatedat,s.deletedat,s.userid,s.topic,s.delid,s.recvseqid,
+		s.readseqid,s.modewant,s.modegiven,u.public,s.private
+		FROM subscriptions AS s JOIN users AS u ON s.userid=u.id
+		WHERE s.topic=?`
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+		if !opts.User.IsZero() {
+			// For p2p topics we have to fetch both users otherwise public cannot be swapped.
+			if tcat != t.TopicCatP2P {
+				q += " AND s.userid=?"
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+// OwnTopics loads a slice of topic names where the user is the owner.
+func (a *adapter) OwnTopics(uid t.Uid, opts *t.QueryOpt) ([]string, error) {
+	rows, err := a.db.Queryx("SELECT name FROM topics WHERE owner=?", store.DecodeUid(uid))
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	}()
+
+	if hard {
+		if _, err = tx.Exec("DELETE FROM subscriptions WHERE topic=?", topic); err != nil {
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+			return err
+		}
+
+		if _, err = tx.Exec("DELETE FROM topictags WHERE topic=?", topic); err != nil {
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+			return err
+		}
+
+		if _, err = tx.Exec("DELETE FROM topics WHERE name=?", topic); err != nil {
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+		}
+	} else {
+		now := t.TimeNow()
+		if _, err = tx.Exec("UPDATE subscriptions SET updatedat=?,deletedat=? WHERE topic=?", now, now, topic); err != nil {
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+			return err
+		}
+
+		if _, err = tx.Exec("UPDATE topics SET updatedat=?,deletedat=? WHERE name=?", now, now, topic); err != nil {
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+}
+
+func (a *adapter) TopicUpdateOnMessage(topic string, msg *t.Message) error {
+	_, err := a.db.Exec("UPDATE topics SET seqid=?,touchedat=? WHERE name=?", msg.SeqId, msg.CreatedAt, topic)
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+	cols, args := updateByMap(update)
+	args = append(args, topic)
+	_, err = tx.Exec("UPDATE topics SET "+strings.Join(cols, ",")+" WHERE name=?", args...)
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	// Tags are also stored in a separate table
+	if tags := extractTags(update); tags != nil {
+		// First delete all user tags
+		_, err = tx.Exec("DELETE FROM topictags WHERE topic=?", topic)
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+}
+
+func (a *adapter) TopicOwnerChange(topic string, newOwner, oldOwner t.Uid) error {
+	_, err := a.db.Exec("UPDATE topics SET owner=? WHERE name=?", store.DecodeUid(newOwner), topic)
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+func (a *adapter) SubscriptionGet(topic string, user t.Uid) (*t.Subscription, error) {
+	var sub t.Subscription
+	err := a.db.Get(&sub, `SELECT createdat,updatedat,deletedat,userid AS user,topic,delid,recvseqid,
+		readseqid,modewant,modegiven,private FROM subscriptions WHERE topic=? AND userid=?`,
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+// Update time when the user was last attached to the topic
+func (a *adapter) SubsLastSeen(topic string, user t.Uid, lastSeen map[string]time.Time) error {
+	_, err := a.db.Exec("UPDATE subscriptions SET lastseen=?,useragent=? WHERE topic=? AND userid=?",
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+// TODO: this is used only for presence notifications, no need to load Private either.
+func (a *adapter) SubsForUser(forUser t.Uid, keepDeleted bool, opts *t.QueryOpt) ([]t.Subscription, error) {
+	q := `SELECT createdat,updatedat,deletedat,userid AS user,topic,delid,recvseqid,
+		readseqid,modewant,modegiven,private FROM subscriptions WHERE userid=?`
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+// the latter does not.
+func (a *adapter) SubsForTopic(topic string, keepDeleted bool, opts *t.QueryOpt) ([]t.Subscription, error) {
+	q := `SELECT createdat,updatedat,deletedat,userid AS user,topic,delid,recvseqid,
+		readseqid,modewant,modegiven,private FROM subscriptions WHERE topic=?`
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	}()
+
+	cols, args := updateByMap(update)
+	q := "UPDATE subscriptions SET " + strings.Join(cols, ",") + " WHERE topic=?"
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+// SubsDelete marks subscription as deleted.
+func (a *adapter) SubsDelete(topic string, user t.Uid) error {
+	now := t.TimeNow()
+	res, err := a.db.Exec("UPDATE subscriptions SET updatedat=?, deletedat=? WHERE topic=? AND userid=? AND deletedat IS NULL",
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+func (a *adapter) SubsDelForTopic(topic string, hard bool) error {
+	var err error
+	if hard {
+		_, err = a.db.Exec("DELETE FROM subscriptions WHERE topic=?", topic)
```
`?` vs `$1`

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+		_, err = a.db.Exec("DELETE FROM subscriptions WHERE topic=?", topic)
+	} else {
+		now := t.TimeNow()
+		_, err = a.db.Exec("UPDATE subscriptions SET updatedat=?, deletedat=? WHERE topic=? AND deletedat IS NULL",
```
`?` vs `$1`

There are more cases below where it applies.

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+// CredFail increments failure count of the given validation method.
+func (a *adapter) CredFail(uid t.Uid, method string) error {
+	_, err := a.db.Exec("UPDATE credentials SET updatedat=?
```
Likewise, you can cast it to `*pq.Error` and check for specific error code.

or-else replied:
Method signature has changed. Please update.

or-else replied:
Likewise, you can cast it to *pq.Error and check for specific error code.

or-else commented on tinode-db/tinode.conf:
```diff
@@ -9,7 +9,11 @@
 				"addresses": "localhost:28015"
-			}
+			},
+      "pg": {
```
Please use tabs instead of spaces.

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+	return tx.Commit()
+}
+
+func (a *adapter) UpgradeDb() error {
```
All you need here is `return nil`. All the code below is unnecessary. Postgress adapter does not exist for older versions and consequently does not need to support the upgrades.

A9u replied:
Thanks. I will update

or-else commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2695 @@
+
+// CredFail increments failure count of the given validation method.
+func (a *adapter) CredFail(uid t.Uid, method string) error {
+	_, err := a.db.Exec("UPDATE credentials SET updatedat=?
```
Method signature has changed.

or-else replied:
Github is buggy. It attaches comments to wrong lines. This comments is about 
`store.RegisterAdapter(adapterName, &adapter{})`

A9u commented on server/db/sql/adapter.go:
```diff
@@ -0,0 +1,2624 @@
+package sql
```
Sorry, I will remove it.

A9u commented:
@or-else Thanks for the review comments.  I will update those according to the comments. 

I am working on the query part. I will let you know when it is done.

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	fmt.Println(result)
+	a.db.Close()
+
+	dsn := strings.Replace(a.dsn, "?", "/"+a.dbName+"?", 1)
```
Why is this needed? Wouldn't `a.dsn` already contain the database name?

or-else replied:
This is ok, because db doesn't initially exist.

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	fmt.Println(result)
+
+	if _, err = tx.Exec(
+		`CREATE TABLE users(
```
Instead of executing these "create table" statements separately,
would it be easier to keep the schema separately and apply it in one go?
Check the example in https://github.com/jmoiron/sqlx#usage

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	if _, err = tx.Exec("INSERT INTO kvmeta(key, value) VALUES ('version'," + strconv.Itoa(adpVersion) + ")"); err != nil {
+		return err
+	}
+
```
You may want to create the system topic here, as well.

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		args = append(args, tag)
+	}
+
+	query, args, _ := sqlx.In("DELETE FROM "+table+" WHERE "+keyName+"=? AND tag IN (?)", keyVal, args)
```
Please use postgres syntax for the params here.

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		// Fetch grp & p2p topics
+		q, topq, _ := sqlx.In(
+			"SELECT createdat,updatedat,deletedat,touchedat,name AS id,access,seqid,delid,public,tags "+
+				"FROM topics WHERE name IN (?)", topq)
```
Postgres syntax for params.

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		// Those unmodified will be stripped of Public & Private.
+
+		if opts.Topic != "" {
+			q += " AND topic=?"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+			limit = opts.Limit
+		}
+	}
+	q += " LIMIT ?"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		// Those unmodified will be stripped of Public & Private.
+
+		if !opts.User.IsZero() {
+			q += " AND userid=?"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		}
+	}
+
+	q += " LIMIT ?"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	args = append(args, topic)
+	if !user.IsZero() {
+		// Update just one topic subscription
+		q += " AND userid=?"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+func subsDelForUser(tx *sqlx.Tx, user t.Uid, hard bool) error {
+	var err error
+	if hard {
+		_, err = tx.Exec("DELETE FROM subscriptions WHERE userid=?", store.DecodeUid(user))
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		_, err = tx.Exec("DELETE FROM subscriptions WHERE userid=?", store.DecodeUid(user))
+	} else {
+		now := t.TimeNow()
+		_, err = tx.Exec("UPDATE subscriptions SET updatedat=?, deletedat=? WHERE userid=?",
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+
+	query := "SELECT u.id,u.createdat,u.updatedat,u.access,u.public,u.tags,COUNT(*) AS matches " +
+		"FROM users AS u LEFT JOIN usertags AS t ON t.userid=u.id " +
+		"WHERE t.tag IN (?" + strings.Repeat(",?", len(req)+len(opt)-1) + ") AND u.deletedat IS NULL " +
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		"WHERE t.tag IN (?" + strings.Repeat(",?", len(req)+len(opt)-1) + ") AND u.deletedat IS NULL " +
+		"GROUP BY u.id,u.createdat,u.updatedat,u.public,u.tags "
+	if len(req) > 0 {
+		query += "HAVING COUNT(t.tag IN (?" + strings.Repeat(",?", len(req)-1) + ") OR NULL)>=? "
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		}
+		args = append(args, len(req))
+	}
+	query += "ORDER BY matches DESC LIMIT ?"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+
+	query := "SELECT t.name AS topic,t.createdat,t.updatedat,t.access,t.public,t.tags,COUNT(*) AS matches " +
+		"FROM topics AS t LEFT JOIN topictags AS tt ON t.name=tt.topic " +
+		"WHERE tt.tag IN (?" + strings.Repeat(",?", len(req)+len(opt)-1) + ") AND t.deletedat IS NULL " +
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		"WHERE tt.tag IN (?" + strings.Repeat(",?", len(req)+len(opt)-1) + ") AND t.deletedat IS NULL " +
+		"GROUP BY t.name,t.createdat,t.updatedat,t.public,t.tags "
+	if len(req) > 0 {
+		query += "HAVING COUNT(tt.tag IN (?" + strings.Repeat(",?", len(req)-1) + ") OR NULL)>=? "
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		}
+		args = append(args, len(req))
+	}
+	query += "ORDER BY matches DESC LIMIT ?"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+// Messages
+func (a *adapter) MessageSave(msg *t.Message) error {
+	res, err := a.db.Exec(
+		"INSERT INTO messages(createdAt,updatedAt,seqid,topic,`from`,head,content) VALUES(?,?,?,?,?,?,?)",
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+			" FROM messages AS m LEFT JOIN dellog AS d"+
+			" ON d.topic=m.topic AND m.seqid BETWEEN d.low AND d.hi AND d.deletedfor=?"+
+			" WHERE m.delid=0 AND m.topic=? AND m.seqid BETWEEN ? AND ? AND d.deletedfor IS NULL"+
+			" ORDER BY m.seqid DESC LIMIT ?",
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	}
+
+	// Fetch log of deletions
+	rows, err := a.db.Queryx("SELECT topic,deletedfor,delid,low,hi FROM dellog WHERE topic=? AND delid BETWEEN ? AND ?"+
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	var err error
+	if toDel == nil {
+		// Whole topic is being deleted, thus also deleting all messages.
+		_, err = tx.Exec("DELETE FROM dellog WHERE topic=?", topic)
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+		forUser := decodeUidString(toDel.DeletedFor)
+		var insert *sql.Stmt
+		if insert, err = tx.Prepare(
+			"INSERT INTO dellog(topic,deletedfor,delid,low,hi) VALUES(?,?,?,?,?)"); err != nil {
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+					}
+				}
+
+				where += "m.seqid IN (?" + strings.Repeat(",?", seqCount-1) + ")"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	var values []string
+	strNow := t.TimeNow().Format("2006-01-02T15:04:05.999")
+	// createdat,fileid,msgid
+	val := "VALUES('" + strNow + "',?," + strconv.FormatInt(int64(msgId), 10) + ")"
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+	}
+
+	// Actually add/update DeviceId for the new user
+	_, err = tx.Exec("INSERT INTO devices(userid, hash, deviceId, platform, lastseen, lang) VALUES(?,?,?,?,?,?)",
```
`?` vs `$1`

aforge commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+
+// CredFail increments failure count of the given validation method.
+func (a *adapter) CredFail(uid t.Uid, method string) error {
+	_, err := a.db.Exec("UPDATE credentia
```
`?` vs `$1`

aforge replied:
`?` vs `$1`

aforge replied:
`?` vs `$1`

A9u commented on server/db/postgres/adapter.go:
```diff
@@ -0,0 +1,2696 @@
+// IsOpen returns true if connection to database has been established. It does not check if
+// connection is actually live.
+func (a *adapter) IsOpen() bool {
+	fmt.Println("inside adapter isOpen()")
```
> Out of curiosity: what kind of testing did you give this code so far?

I am able to connect to postgres db and insert user. It is currently failing at insertion of subscription when private field is nil.

or-else commented:
@A9u 

What are your plans for this PR? If you plan to continue, what is your approximate schedule? 

Just having it open may preclude other people from working on the feature.

A9u commented:
> @A9u
> 
> What are your plans for this PR? If you plan to continue, what is your approximate schedule?
> 
> Just having it open may preclude other people from working on the feature.

@or-else 

First of all, I apologize for dragging it on for such a long time. 

 I am still excited and willing to contribute, however the harsh reality is that I am employed full time in a service industry and get very little for personal contributions. To add to that, recently I have met with a serious accident, due to which I am unable to code for ~20 more days.

While I understand it's asking for too much, I would highly appreciate if you could give me some more time (~1.5 months) to raise this PR.

Having said that, if someone else is willing to work on this feature in the meantime, I will withdraw myself in the best interest of the this great project.

No what how things progress moving forward, I thank you for this amazing project.

or-else commented:
@A9u Thanks for the update. Sorry to hear about the accident, I hope you are OK now. 
No worries, the PR can wait. 
Thanks!

A9u commented:
Hi @or-else 
I am getting following error when running the database initializer. 
```pq: syntax error at or near "=?"```

In the gendb.go file, we set Address book as Private field. 
It is an update query of subscriptions table where we set the private field . The private field is of json type. Can you please help me here with the query.

or-else commented:
Postgressql uses `$1` syntax rather than `?`.

A9u commented:
@or-else  When I try to login using the sample data, I get the following error - 
```Scan error on column index 8, name "modewant": AccessMode: invalid character ' '```

The datatype of modewant in database is char(8) which is blank padded. The blank spaces are causing the issue. Can you please tell me how do I resolve it?

or-else commented:
`JSON` or `VARCHAR`?

A9u commented:
> `JSON` or `VARCHAR`?

In case of ```VARCHAR```, I get following error from Scan function in store/types/types.go. 

```Scan error on column index 8, name "modewant": scan failed: data is not a byte slice```

If we use json, then we need to store it as array.

or-else commented:
If you need to store it as bytes then store as `BYTEA`.

I don't understand this: `If we use json, then we need to store it as array.`.

or-else commented:
Please feel free to open a new request of you decide to contribute.
